components/NavigationBar/src/MDCNavigationBar.m

```diff
@@ -466,19 +466,46 @@ - (CGRect)mdc_frameAlignedVertically:(CGRect)frame
 - (CGRect)mdc_frameAlignedHorizontally:(CGRect)frame
                              alignment:(MDCNavigationBarTitleAlignment)alignment {
   switch (alignment) {
-    // Center align title if desired unless leading icons will overlap frame
+    // Center align title
     case MDCNavigationBarTitleAlignmentCenter: {
-      CGFloat xOrigin = CGRectGetMaxX(self.bounds) / 2 - CGRectGetWidth(frame) / 2;
-      CGFloat minX = CGRectGetMinX(frame);
-      // If RTL, we must use distance from maxX to bounds instead
-      if (self.mdc_effectiveUserInterfaceLayoutDirection ==
-          UIUserInterfaceLayoutDirectionRightToLeft) {
-        minX = CGRectGetMaxX(self.bounds) - CGRectGetMaxX(frame);
+      BOOL isRTL = [self mdc_effectiveUserInterfaceLayoutDirection] ==
+                   UIUserInterfaceLayoutDirectionRightToLeft;
+
+      MDCButtonBar *leftButtonBar = self.leadingButtonBar;
+      MDCButtonBar *rightButtonBar = self.trailingButtonBar;
+      UIEdgeInsets textInsets =
+          [[UIDevice currentDevice] userInterfaceIdiom] == UIUserInterfaceIdiomPad ? kTextPadInsets
+                                                                                   : kTextInsets;
+      CGFloat titleLeftInset = textInsets.left;
+      CGFloat titleRightInset = textInsets.right;
+
+      if (isRTL) {
+        leftButtonBar = self.trailingButtonBar;
+        rightButtonBar = self.leadingButtonBar;
+        titleLeftInset = textInsets.right;
+        titleRightInset = textInsets.left;
       }
-      if (minX <= xOrigin) {
+
+      CGFloat midX = CGRectGetMidX(self.bounds);
+      CGFloat leftMidSpaceX = midX - CGRectGetMaxX(leftButtonBar.frame) - titleLeftInset;
+      CGFloat rightMidSpaceX = CGRectGetMinX(rightButtonBar.frame) - midX - titleRightInset;
+      CGFloat halfFrameWidth = CGRectGetWidth(frame) / 2;
+      if (leftMidSpaceX >= halfFrameWidth && rightMidSpaceX >= halfFrameWidth) {
```

Comment:
This is a bit tricky to follow.
Can we add a one line comment above this if?  Something along the lines of the following, which is a guess:
// If left button bar and right button bar will overlap the title shrink the title's frame